### ISSUE TITLE
Enhancement: Forbid empty type argument list

### ISSUE THREAD

**Issue by fisker** [typescript-eslint/typescript-eslint#11561]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

```ts
f1<>();

new f2<>();

function f3<>() { }

class f4 {
  constructor<>() { }
}

const f5 = function <>() { }

interface f6<> {
  test<>();
}

class f7<> {
  test<>() { }
}
```

should be invalid.

### Additional Info

Tests copied from Prettier repo https://github.com/prettier/prettier/blob/08f687d4b2a3a789ae93226025915c1b272b8b8c/tests/format/typescript/error-recovery/generic.ts#L1C1-L19C2

[Playground](https://typescript-eslint.io/play#ts=5.3.3&showAST=es&fileType=.tsx&code=GYRgPAfAFAlA3AWAFDIHYFMDuACYAmSWRFJYAV1QGMAXASwHtVcBmQmbAb2wF9llKANgEMAziNwAWTsmzZKjEdQBOZGvSVtOPZLxLzUi3AFZsAXlwUaDJpq67ktVNXRLgQyulwA2SNKSznRTZieyRBUXFgAHZfDhlsQOpbbSRuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
